effectiveOfficeBackend/src/main/kotlin/office/effective/features/workspace/repository/WorkspaceRepository.kt

```diff
@@ -0,0 +1,139 @@
+package office.effective.features.workspace.repository
+
+import office.effective.common.exception.UtilityNotFoundException
+import office.effective.common.exception.WorkspaceNotFoundException
+import office.effective.common.exception.WorkspaceTagNotFoundException
+import office.effective.features.workspace.converters.WorkspaceRepositoryConverter
+import office.effective.model.Utility
+import office.effective.model.Workspace
+import org.koin.java.KoinJavaComponent.inject
+import org.ktorm.database.Database
+import org.ktorm.dsl.*
+import org.ktorm.entity.*
+import org.ktorm.support.postgresql.insertOrUpdate
+import java.util.UUID
+
+class WorkspaceRepository {
+    private val database by inject<Database>(clazz = Database::class.java)
+    private val converter by inject<WorkspaceRepositoryConverter>(
+        clazz = WorkspaceRepositoryConverter::class.java
+    )
+
+    /**
+     * Returns whether a workspace with the given id exists
+     */
+    fun workspaceExistsById(workspaceId: UUID): Boolean {
+        return database.workspaces.count { it.id eq workspaceId } > 0
+    }
+
+    /**
+     * Returns whether a utility with the given id exists
+     */
+    fun utilityExistsById(utilityId: UUID): Boolean {
+        return database.utilities.count { it.id eq utilityId } > 0
+    }
+
+    /**
+     * Returns all workspace utilities by workspace id
+     *
+     * Throws WorkspaceNotFoundException if workspace with given id doesn't exist in the database
+     */
+    private fun findUtilitiesByWorkspaceId(workspaceId: UUID): List<Utility> {
+        if (!workspaceExistsById(workspaceId)) {
+            throw WorkspaceNotFoundException("Workspace with id $workspaceId not found")
+        }
+        val modelList = database
+            .from(WorkspaceUtilities)
+            .innerJoin(right = Utilities, on = WorkspaceUtilities.utilityId eq Utilities.id)
+            .select()
+            .where { WorkspaceUtilities.workspaceId eq workspaceId }
+            .map { row ->
+                converter.utilityEntityToModel(
+                    Utilities.createEntity(row), row[WorkspaceUtilities.count]?:0
+                )
+            }
+        return modelList
+    }
+
+    /**
+     * Retrieves a workspace model by its id
+     */
+    fun findById(workspaceId: UUID): Workspace? {
+        val entity: WorkspaceEntity? = database.workspaces.find { it.id eq workspaceId }
+        val utilities: List<Utility> = findUtilitiesByWorkspaceId(workspaceId)
+        return entity?.let { converter.entityToModel(it, utilities) }
```

Comment:
Perhaps we should throw an error if nothing is found?

Reply:
No, in all methods repository should just find all the data it can find. The result is processed by service.